tensorflow/core/grappler/optimizers/data/filter_with_random_uniform_fusion.cc

```diff
@@ -0,0 +1,206 @@
+  for (const NodeDef& func_node : function.node_def()) {
+    if (func_node.op() != op) {
+      continue;
+    }
+    if (func_node.name() + match != func) {
+      continue;
+    }
+    return &func_node;
+  }
+  return NULL;
+}
+
+// This optimization fuse one of the following two forms of
+// filter + random_uniform predication into a single data sampling operation:
+// fuse:
+//   filter
+//   |
+//   + predication: less [0]
+//                  |
+//                  + random_uniform [1]
+//                  |
+//                  + rate
+// or:
+//   filter
+//   |
+//   + predication: less
+//                  |
+//                  + random_uniform[]
+//                  |
+//                  + rate
+// into:
+//   sampling(rate)
+Status FilterWithRandomUniformFusion::OptimizeAndCollectStats(Cluster* cluster,
+                                                  const GrapplerItem& item,
+                                                  GraphDef* output,
+                                                  OptimizationStats* stats) {
+  *output = item.graph;
+  MutableGraphView graph(output);
+  absl::flat_hash_set<string> nodes_to_delete;
+  float rate;
+
+  for (const NodeDef& node : item.graph.node()) {
+    // stage 1 -- recognition
+    if (node.op() != "FilterDataset") {
+      continue;
+    }
+
+    // Use a more descriptive variable name
+    const NodeDef& filter_node = node;
+
+    // find predicate function of the node
+    const auto& predicate = filter_node.attr().at("predicate");
+    const string func_name = predicate.func().name();
+
+    bool function_match = false;
+    // find the function that matches func_name
+    for (const auto& function : item.graph.library().function()) {
+      if (function.signature().name() == func_name) {
+        if (function.ret().size() != 1) {
+          continue;
+        }
+        auto it = function.ret().begin();
+        string node_name = it->second;
+        const NodeDef* less_node;
+        const NodeDef* func_node =
+            FunctionFindNodeDef(function, "Identity", node_name, ":output:0");
+        if (func_node != NULL) {
+          node_name = func_node->input(0);
+        }
+        func_node = FunctionFindNodeDef(function, "StridedSlice", node_name,
+                                        ":output:0");
+        if (func_node != NULL) {
+          // for form one: datasetS = datasetS.filter(lambda x:
+          // tf.less(tf.random_uniform([1]), rate)[0])
+          less_node = FunctionFindNodeDef(function, "Less", func_node->input(0),
+                                          ":z:0");
+        } else {
+          // for form two: datasetS = datasetS.filter(lambda _:
+          // tf.random_uniform([]) < rate)
+          less_node = FunctionFindNodeDef(function, "Less", node_name, ":z:0");
+        }
+        if (less_node == NULL) {
+          continue;
+        }
+
+        // check whether the function is actually doing
+        // random_uniform[0.0, 1.0) < rate
```

Comment:
where do you check the internal boundaries of the `random_uniform` op?

Reply:
check for boundaries had been added.  The original implementation only accept the default form of random_uniform, which boundary is [0.0, 1.0).  The updated implementation also accept the parameter form which compute the output of random_uniform with boundary value.   The updated implementation would check the parameter form and see whether the boundary value is still [0.0, 1.0).  This will make this optimization more adaptive.
See 

https://github.com/tensorflow/tensorflow/pull/25384/commits/fa11069c3e85ffbcae19994bdba09c30eee0a6de#diff-ccdaa4724ea485e41d03d236b6ee46d0R202

https://github.com/tensorflow/tensorflow/pull/25384/commits/fa11069c3e85ffbcae19994bdba09c30eee0a6de#diff-ccdaa4724ea485e41d03d236b6ee46d0R214

https://github.com/tensorflow/tensorflow/pull/25384/commits/fa11069c3e85ffbcae19994bdba09c30eee0a6de#diff-ccdaa4724ea485e41d03d236b6ee46d0R220